url647b55da7c_Nikhilkumar13_emuSr_tgz-pJ8-MainJ8/src/backend/StepRun.java

```diff
@@ -0,0 +1,393 @@
+                     //end_pos=initial_pos+strLine.length();
+                    end_pos = ScanFile.pos.intValue();
+                    //System.out.println("end pos " + end_pos + "initialpos " + initial_pos);
+                    frontend.FrontEnd.activepane.getHighlighter().addHighlight(initial_pos, end_pos, frontend.handlers.cyanPainter);
+                    backend.Register.r[16].b = ScanFile.pos.intValue();//update the prgram counter
+                    ScanFile.current_line = ScanFile.getLineNumber(frontend.FrontEnd.activepane, initial_pos);
+                    //System.out.println("\n" + strLine + " at line" + String.valueOf(backend.ScanFile.current_line));
+                    // System.out.println("commentflag in step run " + commentflag);
+
+                    RemoveComments.setflag(commentflag);
+                    strLine = backend.RemoveComments.RC(strLine.trim());
+                    commentflag = RemoveComments.getcommetnflag();
+
+                    if (strLine.isEmpty())
+                    {
+                        return;
+
+                    } 
+                      frontend.FrontEnd.highlight(initial_pos,strLine, Color.CYAN);
+                    strLine=strLine.toUpperCase();
+                   if (strLine.contains(":"))
+                    {
+
+                        StringTokenizer st2 = new StringTokenizer(strLine, ":");
+                        int ct_label = st2.countTokens();
+                        if (ct_label == 1)
+                        {
+                            //System.err.println("label in step run");
+
+                            return;
+                        } else
+                        {
+                            if (ct_label == 2)
+                            {
+                                strLine = st2.nextToken();
+                                strLine = st2.nextToken();
+
+                            }
+                            strLine = strLine.trim();
+
+                            ScanFile.tokenizeInstruction(strLine);
+                            if (frontend.FrontEnd.exceptionraised == 0)
+                            {
+                                FrontEnd.appendToPane(FrontEnd.statuswindow,"Instruction ",Color.GREEN);
+                                    FrontEnd.appendToPane(FrontEnd.statuswindow,ScanFile.stepins,Color.BLACK);
+                                    FrontEnd.appendToPane(FrontEnd.statuswindow, " Executed Successfully \n",Color.GREEN);
+                            } else
+                            {
+                                FrontEnd.appendToPane(FrontEnd.statuswindow,"Instruction ",Color.RED);
+                                    FrontEnd.appendToPane(FrontEnd.statuswindow,ScanFile.stepins,Color.BLACK);
+                                    FrontEnd.appendToPane(FrontEnd.statuswindow, " Failed To Execute \n",Color.RED);
+                            }
+                        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 141:162, 307:334.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.